common/bip322/signature.go

```diff
@@ -0,0 +1,141 @@
+package bip322
+
+import (
+	"bytes"
+	"encoding/base64"
+
+	"github.com/btcsuite/btcd/txscript"
+	"github.com/btcsuite/btcd/wire"
+)
+
+// Signature is the signed and extracted toSign transaction
+type Signature wire.MsgTx
+
+func DecodeSignature(b64 string) (*Signature, error) {
+	decoded, err := base64.StdEncoding.DecodeString(b64)
+	if err != nil {
+		return nil, err
+	}
+
+	tx := wire.NewMsgTx(wire.TxVersion)
+	if err := tx.Deserialize(bytes.NewReader(decoded)); err != nil {
+		return nil, err
+	}
+
+	return (*Signature)(tx), nil
+}
+
+// Encode encodes the tx to a base64 string
+func (s Signature) Encode() (string, error) {
+	var buf bytes.Buffer
+
+	tx := wire.MsgTx(s)
+
+	if err := tx.Serialize(&buf); err != nil {
+		return "", err
+	}
+	return base64.StdEncoding.EncodeToString(buf.Bytes()), nil
+}
+
+// Verify validates the BIP0322 full proof of funds
+// our version does not check the input sequences in order to be compatible with offchain transactions
+func (s Signature) Verify(message string, prevoutFetcher txscript.PrevOutputFetcher) error {
+	if len(s.TxIn) < 2 {
+		return ErrInvalidTxNumberOfInputs
+	}
+
+	if len(s.TxOut) == 0 {
+		return ErrInvalidTxNumberOfOutputs
+	}
+
+	// the first input of the tx is always the toSpend tx,
+	// we use the input index 1 to get initial pkscript use to craft toSpend
+	secondInputPrevout := prevoutFetcher.FetchPrevOutput(s.TxIn[1].PreviousOutPoint)
+	if secondInputPrevout == nil {
+		return ErrPrevoutNotFound
+	}
+
+	// craft the toSpend tx
+	toSpend := craftToSpendTx(message, secondInputPrevout.PkScript)
+	toSpendHash := toSpend.TxHash()
+
+	// overwrite the prevoutFetcher to include the toSpend tx
+	prevoutFetcher = &bip322PrevoutFetcher{
+		prevoutFetcher: prevoutFetcher,
+		toSpend:        toSpend,
+	}
+
+	// verify that toSpend tx is used as first input
+	if !s.TxIn[0].PreviousOutPoint.Hash.IsEqual(&toSpendHash) {
+		return ErrInvalidTxWrongTxHash
+	}
+	if s.TxIn[0].PreviousOutPoint.Index != 0 {
+		return ErrInvalidTxWrongOutputIndex
+	}
+
+	tx := wire.MsgTx(s)
+
+	txSigHashes := txscript.NewTxSigHashes(&tx, prevoutFetcher)
+	sigCache := txscript.NewSigCache(1000)
+
+	for i, input := range s.TxIn {
+		prevout := prevoutFetcher.FetchPrevOutput(input.PreviousOutPoint)
+		if prevout == nil {
+			return ErrPrevoutNotFound
+		}
+
+		engine, err := txscript.NewEngine(
+			prevout.PkScript,
+			&tx,
+			i,
+			txscript.StandardVerifyFlags,
+			sigCache,
+			txSigHashes,
+			prevout.Value,
+			prevoutFetcher,
+		)
+		if err != nil {
+			return err
+		}
+
+		if err := engine.Execute(); err != nil {
+			return err
+		}
+
+	}
+
+	return nil
+}
+
+func (s *Signature) GetOutpoints() []wire.OutPoint {
+	outpoints := make([]wire.OutPoint, 0, len(s.TxIn)-1)
+	for _, input := range s.TxIn[1:] {
+		outpoints = append(outpoints, input.PreviousOutPoint)
+	}
+	return outpoints
+}
+
+func (s *Signature) ContainsOutputs() bool {
+	if len(s.TxOut) > 0 {
```

Comment:
if no outputs= false
if 1 outputs and out[0] ==opReturnPkScript = false
else true